### PR TITLE
Add example for stable diffusion 1 5 model

### LINKED ISSUE #5042: Add demo for stable diffusion 1.5 model
stable diffusion 1.5 model is running e2e so we need to add example for the model

### PR DESCRIPTION
**Pull request by ctr-lelanchelianTT**
### Ticket
closes https://github.com/tenstorrent/tt-xla/issues/5042

### Problem description
Stable diffusion 1 5 model is passing e2e so we need to add example for this model 

### What's changed
- Adds `examples/pytorch/sd_v1_5_pipeline.py` for Stable Diffusion 1.5 (`stable-diffusion-v1-5/stable-diffusion-v1-5`), mirroring the existing `sd_v1_4_pipeline.py`.
- UNet runs on the TT backend; VAE + PNDM scheduler on CPU; CFG denoising loop; 512x512 output.

### Checklist
- [x] New/Existing tests provide coverage for changes

I have attached the log for your reference:
[sd_v1_5.log](https://github.com/user-attachments/files/28495121/sd_v1_5.log)

output image:
test_sd15_output

### CODE REVIEW

**Review by meenakshiramanathan1 — APPROVED** [review]

File: examples/pytorch/sd_v1_5_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+    """Pipeline for text-to-image generation with Stable Diffusion 1.5."""
+
+    def __init__(self, config: SD15Config):
+        self.config = config
+        self.device = config.device
+        self.model_id = config.model_id
+        self.latents_width = config.latents_width
+        self.latents_height = config.latents_height
+        self.vae_on_tt = config.vae_on_tt
+        self.clip_on_tt = config.clip_on_tt
```
Comment (vkovacevicTT) [inline examples/pytorch/sd_v1_5_pipeline.py]: Why do we unpack these when we have them all in config?

File: examples/pytorch/sd_v1_5_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+    def load_models(self):
+        self.vae = AutoencoderKL.from_pretrained(
+            self.model_id,
+            subfolder="vae",
+            torch_dtype=torch.float32,
+            device_map=self.device,
+        )
+        if self.vae_on_tt:
+            self.vae.compile(backend="tt")
+            self.vae = self.vae.to(xm.xla_device())
```
Comment (vkovacevicTT) [inline examples/pytorch/sd_v1_5_pipeline.py]: `self.device` here is a bit misleading.
We default it to "cpu", we have `self.vae_on_tt` to signal we want it on TT device, and then we use `xm.xla_device()`, which beats the purpose of having a device attribute.

File: examples/pytorch/sd_v1_5_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+import torch_xla
+import torch_xla.core.xla_model as xm
+import torch_xla.runtime as xr
+from diffusers import AutoencoderKL, PNDMScheduler, UNet2DConditionModel
+from PIL import Image
+from transformers import CLIPTextModel, CLIPTokenizer
+
+
+class SD15Config:
+    def __init__(self, device="cpu", vae_on_tt=False, clip_on_tt=True):
```
Comment (vkovacevicTT) [inline examples/pytorch/sd_v1_5_pipeline.py]: Do we want `clip_on_tt=True` by default?

File: examples/pytorch/sd_v1_5_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+
+            # --- Text encoding (CLIP) ---
+            negative_prompt = negative_prompt or ""
+
+            cond_tokens = self.tokenizer(
+                [prompt], padding="max_length", max_length=77
+            ).input_ids
+            uncond_tokens = self.tokenizer(
+                [negative_prompt], padding="max_length", max_length=77
+            ).input_ids
```
Comment (vkovacevicTT) [inline examples/pytorch/sd_v1_5_pipeline.py]: Does this work when batch_size > 1?

File: examples/pytorch/sd_v1_5_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+        """Generate an image from a text prompt. Returns tensor (B, 3, H, W)."""
+
+        batch_size = 1 if isinstance(prompt, str) else len(prompt)
+
+        tt_cast = lambda x: (
+            x.to(dtype=torch.bfloat16).to(device=xm.xla_device())
+            if x.device == torch.device("cpu")
+            else x.to(dtype=torch.bfloat16)
+        )
+        cpu_cast = lambda x: x.to("cpu").to(dtype=torch.float16)
```
Comment (vkovacevicTT) [inline examples/pytorch/sd_v1_5_pipeline.py]: Do we need this if in `tt_cast`?

File: examples/pytorch/sd_v1_5_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+            # --- Denoising loop (UNet on TT) ---
+            for i, timestep in enumerate(self.scheduler.timesteps):
+
+                model_input = torch.cat([latents] * 2)
+                model_input = self.scheduler.scale_model_input(model_input, timestep)
+
+                # CPU → TT
+                model_input = tt_cast(model_input)
+                timestep_tt = tt_cast(timestep.unsqueeze(0))
+                encoder_hidden_states = tt_cast(encoder_hidden_states)
```
Comment (vkovacevicTT) [inline examples/pytorch/sd_v1_5_pipeline.py]: Do `encoder_hidden_states` change across loops?

**Review by vkovacevicTT — COMMENTED** [review]
I don't mind this being in the `examples/` dir as is, but since two benchmark PRs (https://github.com/tenstorrent/tt-xla/pull/5085, https://github.com/tenstorrent/tt-xla/pull/5044) depend on this implementation I would suggest some changes.

First, there is a lot of code duplication between this PR and https://github.com/tenstorrent/tt-xla/pull/5075 and https://github.com/tenstorrent/tt-xla/pull/5074. 
Let's consider restructuring them and extracting utilities.

Second, we should consider moving the model pipeline implementations to `tt_forge_models`. 
Keep in mind that we do need some flexibility, like passing compile configs e.g. optimization-level specified [here](https://github.com/tenstorrent/tt-xla/pull/5044/changes#diff-7e62f3e7e051e48ba52d7a5d66608661cb4f42a3dc29a91b04ccc0fb0a87fc49R350-R356). 
If moving the whole pipeline to `tt_forge_models` makes things more complicated, we could implement some main model pipeline functionalities (like `generate` function) in the benchmarks (and examples).
I just think it's important we don't import anything from `examples/` dir in the benchmarks.

File: examples/pytorch/sd_v1_5_pipeline.py (hunk 2 above)
Comment (sgligorijevicTT) [reply examples/pytorch/sd_v1_5_pipeline.py]: I 100% agree, please address this

File: examples/pytorch/sd_v1_5_pipeline.py (hunk 6 above)
Comment (sgligorijevicTT) [reply examples/pytorch/sd_v1_5_pipeline.py]: This should be hoisted out of the loop, and perhaps some of these casts can be avoided to make the code cleaner

File: examples/pytorch/sd_v1_5_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+                    timestep_tt,
+                    encoder_hidden_states,
+                ).sample
+
+                # TT → CPU
+                unet_output = cpu_cast(unet_output)
+
+                # CFG blending (CPU)
+                uncond_output, cond_output = unet_output.chunk(2)
+                model_output = uncond_output + (cond_output - uncond_output) * cfg_scale
```
Comment (sgligorijevicTT) [inline examples/pytorch/sd_v1_5_pipeline.py]: Is there a need to do CFG on the CPU?

File: examples/pytorch/sd_v1_5_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+
+                # Scheduler step (CPU)
+                latents = cpu_cast(latents)
+                latents = self.scheduler.step(
+                    model_output, timestep, latents
```
Comment (sgligorijevicTT) [inline examples/pytorch/sd_v1_5_pipeline.py]: For my understanding, why does the scheduler step need to take in latents?
Also, does this have to happen on the CPU?

File: examples/pytorch/sd_v1_5_pipeline.py (hunk 1 above)
Comment (ctr-lelanchelianTT) [reply examples/pytorch/sd_v1_5_pipeline.py]: Restructured: the implementation now lives in `tt_forge_models` (companion tenstorrent/tt-forge-models#720) and this example is a thin wrapper that imports it. In the canonical pipeline I dropped the misleading `device` unpack; the remaining locals are just readability aliases for config fields.

File: examples/pytorch/sd_v1_5_pipeline.py (hunk 2 above)
Comment (ctr-lelanchelianTT) [reply examples/pytorch/sd_v1_5_pipeline.py]: Addressed in the `tt_forge_models` pipeline: removed the `device` config attr entirely and use the explicit `"cpu"` literal for host placement. TT placement stays separate via `.compile(backend="tt")` + `xm.xla_device()`, so there is no longer a misleading "device" attribute.

File: examples/pytorch/sd_v1_5_pipeline.py (hunk 3 above)
Comment (ctr-lelanchelianTT) [reply examples/pytorch/sd_v1_5_pipeline.py]: Good catch — flipped the default to `clip_on_tt=False`. CLIP text embeddings are precision-sensitive and running CLIP on TT in bf16 can wash out the prompt conditioning, so it is now opt-in.

File: examples/pytorch/sd_v1_5_pipeline.py (hunk 4 above)
Comment (ctr-lelanchelianTT) [reply examples/pytorch/sd_v1_5_pipeline.py]: It is written/validated for single-prompt generation only. Rather than silently mishandle batches, I added an explicit `assert batch_size == 1` in the pipeline.

File: examples/pytorch/sd_v1_5_pipeline.py (hunk 5 above)
Comment (ctr-lelanchelianTT) [reply examples/pytorch/sd_v1_5_pipeline.py]: Right — removed the redundant cast; `tt_cast` already handles the dtype + device move.

File: examples/pytorch/sd_v1_5_pipeline.py (hunk 6 above)
Comment (ctr-lelanchelianTT) [reply examples/pytorch/sd_v1_5_pipeline.py]: They do not change across the loop — hoisted the `encoder_hidden_states` cast out of the denoising loop so it runs once before the loop.

File: examples/pytorch/sd_v1_5_pipeline.py (hunk 2 above)
Comment (ctr-lelanchelianTT) [reply examples/pytorch/sd_v1_5_pipeline.py]: Done — removed the `device` attr in the `tt_forge_models` pipeline and use the explicit `"cpu"` literal for host placement.

File: examples/pytorch/sd_v1_5_pipeline.py (hunk 6 above)
Comment (ctr-lelanchelianTT) [reply examples/pytorch/sd_v1_5_pipeline.py]: Hoisted the constant `encoder_hidden_states` cast out of the loop (runs once now) and removed the redundant per-step latents cast.

File: examples/pytorch/sd_v1_5_pipeline.py (hunk 7 above)
Comment (ctr-lelanchelianTT) [reply examples/pytorch/sd_v1_5_pipeline.py]: CFG blending is a cheap elementwise op on the small noise-pred tensors; running it on CPU between the TT UNet forwards avoids round-tripping tiny tensors to the device. It can live on TT but the gain is negligible, so it stays on CPU (added a comment noting this).

File: examples/pytorch/sd_v1_5_pipeline.py (hunk 8 above)
Comment (ctr-lelanchelianTT) [reply examples/pytorch/sd_v1_5_pipeline.py]: The diffusers scheduler (PNDM here) is stateful and runs on CPU; `step(model_output, t, latents)` is the upstream API — it consumes the current latents to produce the next sample. Keeping it on CPU matches upstream numerics exactly; only the heavy UNet forward is offloaded to TT.

**Comment by ctr-lelanchelianTT** [comment]
Thanks for the review  addressed the comments

**Review by sgligorijevicTT — APPROVED** [review]
